### ISSUE TITLE
Socket Connection

### ISSUE THREAD

**Issue by shinetots** [mozilla/libaudit-go#8]
I am using libaudit library and my program runs in ubuntu. Can libaudit creates the netlink connection in CentOS(guest os) running on vbox instead in ubuntu(host os)?

**Comment by arunk-s** [comment]
Can you explain more on exactly what behaviour do you want ?
Do you need to run `libaudit` on the ubuntu (host os) and connect to the CentOS via `netlink` which is running on vbox ? 
or you want to run the `libaudit` on CentOS (running in vbox) which will create a netlink connection in CentOS only ?
Currently the `netlink` connection is opened on the OS kernel (with root permissions) on which `libaudit` is running.
No option is provided to specify the source for the connection as the usual use-case is to retrieve the audit-messages from * the OS on which `libaudit` is running *.
However if you want to connect to a guest OS on vbox, I suppose you can start from https://github.com/mozilla/libaudit-go/blob/2955e1ba66ce4d2fb902400180e4dda200526393/libaudit.go#L105 to change the code accordingly.

**Comment by shinetots** [comment]
Thanks! Yes, I want to connecti libaudit inside CentOS(running in vbox) instead on ubuntu(host). But my program will run on ubuntu. Also, is it possible for libaudit library to tail the audit.log file? Additional changes or new log events should be automatically displayed rather than running the program again.

**Comment by shinetots** [comment]
I tried some of the implementations in your main.go. No output for the events.
https://github.com/mozilla/audit-go/blob/master/main.go#L30
https://github.com/mozilla/audit-go/blob/master/main.go#L156

**Comment by arunk-s** [comment]
As I've explained earlier "currently no option is provided to specify the source for the connection as the usual use-case is to retrieve the audit-messages from the OS on which libaudit is running". So you'll have to change the source code of `libaudit-go` if you want to connect to vbox guest OS via netlink.

Regarding you second question, the purpose of `libaudit-go` is to provide a standalone audit library which _doesn't use `auditd` daemon_ and takes care of configuring audit rules itself.
As the audit.log file is generated by `auditd` daemon, it won't suit the purpose of `libaudit` to just tail `audit.log`. If you want that scenario you can directly use `tail -f  audit.log`.
Also, if you want `audit.log` to be retrievable on your host OS as well and want to tail it , you can just ssh into the guest OS on vbox through your host OS and tail the same file. Plenty of online resources are available which explain how to ssh into vbox OS.

**Comment by shinetots** [comment]
Ok, I got your point. I have tried some implementations in your main.go especially in getting audit events. However, there is no output for getting an audit event. https://github.com/mozilla/audit-go/blob/master/main.go#L156

**Comment by arunk-s** [comment]
In what kind of environment or setup are you trying the implementation ? 
I mean, if you want audit-events to be collected on an OS, the best way is to setup [audit-go](https://github.com/audit-go) and run:
```bash
sudo go run main.go audit.rules
```
where `audit.rules` can have your desired audit rules in similar form as they are required for `auditd` daemon.
Then the `audit-go`program will start showing output for every audit-event.

**Comment by shinetots** [comment]
Thank you. I got you point already. When I ran sudo go run main.go audit.rules, raw events are shown. Is it possible that only parts of the audit event will be shown? For example, for type=SECCOMP, only "exe", "ip","syscall" and "uid" will be displayed? 

Another  question. Is heka plugin will be responsible for decoding contents of the audit events?

**Comment by arunk-s** [comment]
I don't think that should happen. But you can check the same with logs filed in `audit.log` for the same rule as read by `audit-go` by setting them with `auditd`.
The `heka.toml` available at `audit-go` uses an external sandbox decoder `audit_decoder.lua` to parse the content of audit-events and make them available to appropriate sources. Instructions to setup heka to use external decoder are given at readme on the audit-go repo.

**Comment by shinetots** [comment]
Ok. Are the contents of audit-events be decoded into in layman's term? For example, instead of syscall=2, the output would be syscall=open.

**Comment by arunk-s** [comment]
Sorry, I don't think that is possible currently. You'll have to use `ausyscall` utility for that. For a better understanding of audit records I recommend this guide https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/sec-Understanding_Audit_Log_Files.html

**Comment by shinetots** [comment]
Ok, I understand. By the way, for the file watches, I always get this output. Type:EOF Info: {}. The info is always null.

**Comment by arunk-s** [comment]
Oh, I think you should open a new issue regarding that problem and please mention the exact setup and the rules that you were using over there. That will further help us to streamline the problem.

**Comment by arunk-s** [comment]
Closing this, as now the discussion is moved onto #9
